orbit/pkg/table/app_sso_platform/app_sso_platform_darwin.go

```diff
@@ -0,0 +1,247 @@
+	}
+	if deviceConfigJSON == "(null)" {
+		log.Debug().Msg("device not registered")
+		return nil, nil
+	}
+	deviceConfig := struct {
+		DeviceSigningCertificate string `json:"deviceSigningCertificate"`
+		ExtensionIdentifier      string `json:"extensionIdentifier"`
+	}{}
+	if err := json.Unmarshal([]byte(deviceConfigJSON), &deviceConfig); err != nil {
+		return nil, fmt.Errorf("could not unmarshal \"Device Configuration\" JSON: %w", err)
+	}
+	if expectedExtensionIdentifier != deviceConfig.ExtensionIdentifier {
+		log.Debug().Str("extensionIdentifier", deviceConfig.ExtensionIdentifier).Msg("device registered, but found unmatched extension")
+		return nil, nil
+	}
+	dsc, err := base64.RawURLEncoding.DecodeString(deviceConfig.DeviceSigningCertificate)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode \"deviceSigningCertificate\": %w", err)
+	}
+	deviceSigningCertificate, err := x509.ParseCertificate(dsc)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse \"deviceSigningCertificate\": %w", err)
+	}
+	if deviceSigningCertificate.Subject.CommonName == "" {
+		return nil, errors.New("empty subject common name in \"deviceSigningCertificate\"")
+	}
+	log.Debug().Str(
+		"\"Device Configuration\"", deviceSigningCertificate.Subject.CommonName,
+	).Msg("found device ID")
+	userConfig := struct {
+		KerberosStatus []map[string]any `json:"kerberosStatus"`
+	}{}
+	if userConfigJSON == "(null)" {
+		log.Debug().Msg("user not registered")
+		return &appSSOPlatformData{
+			extensionIdentifier: deviceConfig.ExtensionIdentifier,
+			deviceID:            deviceSigningCertificate.Subject.CommonName,
+			realm:               expectedRealm,
+			userPrincipalName:   "",
+		}, nil
+	}
+	if err := json.Unmarshal([]byte(userConfigJSON), &userConfig); err != nil {
+		return nil, fmt.Errorf("could not unmarshal \"User Configuration\" JSON: %w", err)
+	}
+	if len(userConfig.KerberosStatus) == 0 {
+		return nil, errors.New("\"kerberosStatus\" has no entries")
+	}
+	realm_, ok := userConfig.KerberosStatus[0]["realm"]
+	if !ok {
+		return nil, errors.New("missing \"realm\" key in \"kerberosStatus\"")
+	}
+	realm, ok := realm_.(string)
+	if !ok {
+		return nil, fmt.Errorf("unexpected type for \"realm\" key in \"kerberosStatus\": %T", err)
+	}
+	upn_, ok := userConfig.KerberosStatus[0]["upn"]
+	if !ok {
+		return nil, errors.New("missing \"upn\" key in \"kerberosStatus\"")
+	}
+	upn, ok := upn_.(string)
+	if !ok {
+		return nil, fmt.Errorf("unexpected type for \"upn\" key in \"kerberosStatus\": %T", err)
+	}
+	if upn == "" {
+		return nil, errors.New("empty \"upn\" key in \"kerberosStatus\"")
+	}
+	if expectedRealm != realm {
+		log.Debug().Str("realm", realm).Msg("user registered, but found unmatched realm")
+		return &appSSOPlatformData{
+			extensionIdentifier: deviceConfig.ExtensionIdentifier,
+			deviceID:            deviceSigningCertificate.Subject.CommonName,
+			realm:               expectedRealm,
+			userPrincipalName:   "",
+		}, nil
+	}
```

Comment:
I don't follow the logic here.  We're requiring that `realm` be specified in the query, but if we don't get a matching realm in the results, we return a row that has the _expected_ realm, but no principal name?

Reply:
Correct. It's sort of SQL behavior (to return what the requestor sent, it then gets filtered out by osquery/sqlite). What you are saying is: "for the requested realm, the table found an empty user principal name."

Reply:
Discussed more in: https://fleetdm.slack.com/archives/C084F4MKYSJ/p1750433505820709